### PR TITLE
Add Firebase externals for admin & functions

### LINKED ISSUE #19: Add declarations for "firebase"
Pretty popular service
https://firebase.google.com/docs/web/setup

### PR DESCRIPTION
**Pull request by pavlospt**
* Adds JS externals for Firebase SDK version 1.0.0
* Resolves https://github.com/Kotlin/js-externals/issues/19

Since this is my first contribution, please let me know if I miss anything or something is wrong :)

### CODE REVIEW

**Comment by bashor** [comment]
Thank you for the contribution! We'll review it as soon as possible.

**Review by anton-bannykh — CHANGES_REQUESTED** [review]
Thank you for your contribution!

There are some issues though, which prevent us from accepting the PR. Most are with general layout, but that's quite important to us, because publishing scripts rely on it. =)

1. At the moment we don't support nested layouts (e.g. firebase-admin and firebase-functions have to be directly in the externals folder.

1. Version folders correspond to different major versions (e.g. v5 for `firebase-admin`)

1. For NPM package `foo@x.y.z` the `version` in `konfig.json` should be `x.y.0`. In other words we keep major and minor versions and use out own patch version for the potential updates in kotlin external declarations. Also see https://github.com/Kotlin/js-externals#how-to-choose-version-number

1. There should be `README.md` files. Empty will do. Otherwise publishing script fails =( Maybe we should remove this requirement, it's a bit silly =(

1. There should be some minimal tests so that we can check at lease some API is usable (see `jquery/_testShared`). This is mostly a smoke test for different module system usage. Using just a few classes or functions should suffice.

1. Packages. External declarations should reside in `package js.externals.<library-name>`. Otherwise using several at the same time would be a mess.

1. Some type definitions seem to be missing (e.g. `CloudFunction`)?

1. Could you add headers like [these](https://github.com/Kotlin/js-externals/blob/3483c81e1bad279b35b922f7237945b80b52868f/externals/jquery/v2/src/index.kt#L1)? That's required for copyright reasons. Also this is an opportunity to [immortalize](https://github.com/Kotlin/js-externals/blob/3483c81e1bad279b35b922f7237945b80b52868f/externals/paho-mqtt/v3/src/index.kt#L4) your name =)

I hope this list doesn't intimidate you, because most of it is very easily fixable.  =) Also some points aren't listed in the README (sorry!) so there is no way you could have known. =)

Thank you once again for your contribution! I really hope this goes through - there has been some demand for these declarations. =)

**Comment by pavlospt** [comment]
Great! Thank you for your points, will try to fix them ASAP. Will probably ping you for some things that do not look very clear at the time, but I hope i will be able to get along with most of them :)

**Comment by pavlospt** [comment]
I think I covered most of them. Although I have some questions about 3 & 7!

Regarding 3) what should I change? The `konfig.json` is in the requested format. Am I missing something else? :)

Regarding 7) I used `ts2kt` tool to generate the files. Is it possible that it failed to convert a definition or should I look in case I missed a file?

**Comment by anton-bannykh** [comment]
@pavlospt Thanks! Regarding 3) answered in slack. We probably should add that to the README.

7) Could be. Was it  `@firebase/functions-types` you were converting?

**Comment by pavlospt** [comment]
Seems that we have a blocker issue https://github.com/Kotlin/ts2kt/issues/79 

`ts2kt` fails to convert `export declare type` fields :) Would be glad to know what those fields should be converted to, in order to fix the current PR and add what's missing!

**Comment by anton-bannykh** [comment]
Well, it's no exactly a blocker. It should be possible to be fixed by translating that bit by hand.

File: externals/firebase-functions/v1/src/cloud-functions.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    var resource: Resource
+    var params: Json
+    var authType: dynamic /* String /* "ADMIN" */ | String /* "USER" */ | String /* "UNAUTHENTICATED" */ */ get() = definedExternally; set(value) = definedExternally
+    var auth: `T$0`? get() = definedExternally; set(value) = definedExternally
+}
+external open class Change<T>(before: T? = definedExternally /* null */, after: T? = definedExternally /* null */) {
+    open var before: T = definedExternally
+    open var after: T = definedExternally
+    companion object {
+        fun <T> fromObjects(before: T, after: T): `"node_modules/firebase-functions/lib/cloud-functions".Change`<T> = definedExternally
```
Comment (anton-bannykh) [inline externals/firebase-functions/v1/src/cloud-functions.kt]: This doesn't compile

Comment (pavlospt) [reply externals/firebase-functions/v1/src/cloud-functions.kt]: will fix that and other occurrences like it by hand!

File: externals/firebase-admin/v5/src/index.admin.app.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+// 
+// Package: firebase-admin (https://www.npmjs.com/package/firebase-admin)
+//
+// Firebase Admin Node.js SDK is licensed under the Apache License, version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+//
+// Contributed by: Pavlos-Petros Tournaris <p.tournaris@gmail.com>
+
+@file:Suppress("INTERFACE_WITH_SUPERCLASS", "OVERRIDING_FINAL_MEMBER", "RETURN_TYPE_MISMATCH_ON_OVERRIDE", "CONFLICTING_OVERLOADS", "EXTERNAL_DELEGATION", "NESTED_CLASS_IN_EXTERNAL_INTERFACE")
+@file:JsQualifier("admin.app")
+package package js.externals.firebase.admin.app
```
Comment (anton-bannykh) [inline externals/firebase-admin/v5/src/index.admin.app.kt]: ^

File: externals/firebase-admin/v5/src/index.admin.app.kt [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.w3c.fetch.*
+import org.w3c.files.*
+import org.w3c.notifications.*
+import org.w3c.performance.*
+import org.w3c.workers.*
+import org.w3c.xhr.*
+
+external interface App {
+    var name: String
+    var options: admin.AppOptions
```
Comment (anton-bannykh) [inline externals/firebase-admin/v5/src/index.admin.app.kt]: ^ doesn't compile =(
It should have been just `AppOptions`, coupled with an `import js.externals.firebase.admin.AppOptions` at the beginning of the file.

**Review by anton-bannykh — CHANGES_REQUESTED** [review]
It looks much better now, but there are a lot of compilation errors, mostly due to incorrect references to classes (like `admin.AppOptions`.

In order to get the list you could run `./gradlew check`. It will attempt to compile the declarations and perform a few sanity checks.

**Comment by pavlospt** [comment]
@anton-bannykh perfect, thanks for checking it, will resolve the issues and get back :)

**Comment by pavlospt** [comment]
@anton-bannykh Pushed a lot of fixes for the wrong imports. But, now I am stuck on some missing types. Checking the TS typings it seems that it imports `ReadStream/WriteStream` from `fs`, `PromiseLike` is provided by ES5 lib and `Buffer` which I am not sure why it can not find :) Looking forward to your feedback!

**Comment by pavlospt** [comment]
Hey, kind re-ping on this :) Has anyone had the time to take a look :) ?

**Comment by anton-bannykh** [comment]
@pavlospt Sorry, I've been on vacation. =( Will be back to work on Wednesday. Sorry for letting you hanging for so long =(

**Comment by pavlospt** [comment]
@anton-bannykh Ohh sorry did not know! Please take your time :)

File: externals/firebase-admin/v5/src/firestore.FirebaseFirestore.kt [hunk 4]
```diff
@@ -0,0 +1,182 @@
+@file:Suppress("INTERFACE_WITH_SUPERCLASS", "OVERRIDING_FINAL_MEMBER", "RETURN_TYPE_MISMATCH_ON_OVERRIDE", "CONFLICTING_OVERLOADS", "EXTERNAL_DELEGATION", "NESTED_CLASS_IN_EXTERNAL_INTERFACE")
```
Comment (anton-bannykh) [inline externals/firebase-admin/v5/src/firestore.FirebaseFirestore.kt:9]: This file seems to be missing a header comment

**Comment by anton-bannykh** [comment]
It all looks great =)

There is a major issue though =( Fortunately the fix should be simple =)

I've tried writing a simple firebase function, following [this tutorial](https://firebase.google.com/docs/functions/get-started).

When I tried using `initializeApp` function, the resulting JS code for CommonJS module system didn't contain the `require(firebase-functions)`. You can see that by running `./gradlew check` and viewing `externals/firebase-admin/v5/build/classes/kotlin/testCommonJS/v5_testCommonJS.js` file (it contains the compilation result of `test.kt` with module kind "commonjs").

It seems some `@JsModule` annotations are missing. I think `index.admin.kt` and `index.admin.credential.kt` should contain annotations
```
@file:JsNonModule
@file:JsModule("firebase-admin")
```

Something similar should be done for `firebase-functions`.

Thanks!

P.S. Once again, sorry for the long absence.

File: externals/firebase-admin/v5/src/firestore.FirebaseFirestore.kt (hunk 4 above)
Comment (pavlospt) [reply externals/firebase-admin/v5/src/firestore.FirebaseFirestore.kt:9]: Thanks for pointing out!

**Comment by pavlospt** [comment]
@anton-bannykh Thank you for reviewing it, will get the fixes ASAP :)

**Comment by pavlospt** [comment]
@anton-bannykh Fixed those issues and checked the generated test code, that it properly includes the `require('firebase-admin')` & `require('firebase-functions')` declarations, respectively!

**Comment by anton-bannykh** [comment]
Hi! Sorry for the long wait once again =(

So I've been looking at the Firebase documentation. I have doubts it is possible to use the resulting JS code for the WEB use case: https://firebase.google.com/docs/web/setup

It seems the calls to Firebase API are wrong, when compiled with `moduleKind=plain`. Or am I wrong?

P.S. This might be not super-important for the first version. =)
P.P.S. I'm not really familiar with Firebase. Have been meaning to look into it, but things kept coming up =(

**Comment by pavlospt** [comment]
Hello @anton-bannykh, the Firebase converted files are for Node.JS SDK of Firebase Cloud Functions and Admin! Personally, I have tested handwritten Kotlin code and compiled to JS, which perfectly deploys to Cloud Functions! Not sure if the same SDKs are for Web usage.

**Comment by pavlospt** [comment]
Hey @anton-bannykh , it's been a month since the last fixes! Is there any chance that this will get merged? :)

**Comment by anton-bannykh** [comment]
@pavlospt  Sorry, my bad! =( Will merge it asap.

**Comment by pavlospt** [comment]
Hey @anton-bannykh I am really sorry for pinging you again, but it has been 23 days since you said that it will be merged! I am completely sure that you probably have other things to do and forgot that, but I just wanted to see if you have some free time to merge this one :)

**Pull request closed without merging**